### PR TITLE
Explicitly authorize test requests

### PR DESCRIPTION
**Pull request by jaydoane**
Create admin user and use its credentials for requests

Use http requests to create and delete test dbs to avoid potential db
name munging issues downstream

COUCHDB-2897

### CODE REVIEW

File: test/couchdb_mrview_cors_tests.erl [hunk 1]
```diff
@@ -86,14 +95,25 @@ create_db(backdoor, DbName) ->
...
-    ok = fabric:create_db(DbName, [?ADMIN_CTX]).
+    {ok, Status, _, _} = test_request:put(db_url(DbName), [?AUTH], ""),
+    true = lists:member(Status, [201, 202]),
+    ok.
 
 delete_db(backdoor, DbName) ->
     couch_server:delete(DbName, [?ADMIN_CTX]);
 delete_db(clustered, DbName) ->
-    ok = fabric:delete_db(DbName, [?ADMIN_CTX]).
+    {ok, 200, _, _} = test_request:delete(db_url(DbName), [?AUTH]),
```
Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: DELETE /db may eventually return HTTP 202 response.

Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: btw, is there reason to switch from Erlang to HTTP API for this?

**Comment by kxepal** [comment]
LGFM, except on moment, but you seems start to repeat yourself. We have quite a lot of tests where Admin Party implicitly assumed. 

May be it worth to make a single change for test_util to setup cluster without AP and fix all at once?

File: test/couchdb_mrview_cors_tests.erl [hunk 2]
```diff
@@ -86,14 +95,25 @@ create_db(backdoor, DbName) ->
...
+    ok.
 
 
+host_url(PortType) ->
+    Addr = config:get("httpd", "bind_address", "127.0.0.1"),
```
Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: Better to make a switch for httpd/chttpd here as well since you may eventually try to test backdoor interface using chttpd port. Someday.

File: test/couchdb_mrview_cors_tests.erl (hunk 1 above)
Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: In the spirit of YAGNI, I would be inclined to assume 200 for successful DELETE response until otherwise necessary. Also, the docs don't mention 202:
http://docs.couchdb.org/en/latest/api/database/common.html?highlight=delete#delete--db

Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: The reason for the switch from fabric db creation to the HTTP API is because in some downstream cases (specifically to support multiple tenants in one cluster), the DB name gets munged in a way to indicate to which user it belongs. This happens transparently through the HTTP API, but fabric knows nothing about it.

File: test/couchdb_mrview_cors_tests.erl (hunk 2 above)
Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: Again, the principle of YAGNI was what prevented me from making that change for the eventual possibility. However, I'm fine to add it :)

File: test/couchdb_mrview_cors_tests.erl (hunk 1 above)
Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: Documentation is not updated for 2.0 actually, so YAGNI source is in erlang sources for now (:
Ok with fabric to HTTP replacement, though it strange that we do this with Erlang, not JS. Need to revisit this later.

File: test/couchdb_mrview_cors_tests.erl (hunk 2 above)
Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: Thanks! (:

**Comment by jaydoane** [comment]
@kxepal, I hear what you are saying about the Admin Party being assumed for many tests. However, at this point I have not taken a full inventory, and am only focusing on the suites which are failing downstream for various reasons. So far, the ones I've been able to fix by adding explicit authentication are just this one and https://github.com/apache/couchdb-chttpd/pull/91. I'm guessing that the other implicit Admin Party tests may not be doing things (like creating/deleting DBs) which would cause breakage downstream, where an Admin Party is less permissive than in the CouchDB case.

So, in short, I don't think there will be many more PRs such as this. The remaining tests that are failing downstream have much bigger problems, and are likely "unfixable" based on what they are trying to test. I think we'll just have to disable them in downstream builds and move on :(

**Comment by jaydoane** [comment]
Also, I'm wondering what's going on with travis. It's been over 12 hours that the build has been pending...

**Comment by kxepal** [comment]
@jaydoane ): ok
Travis pending is seems because of ASF tasks queue overflow again. This can last for a long time, suddenly.

File: test/couchdb_mrview_cors_tests.erl (hunk 2 above)
Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: Actually, I think the only way the host_url depends on PortType (which can be either 'clustered' or 'backdoor') is expressed via the port/1 function. I'm not sure how you would improve or generalize this.

Or did you mean db_url, which is only valid for clustered PortType? That could be generalized by adding a PortType parameter, but I didn't bother since it's only being used in the clustered cases.

File: test/couchdb_mrview_cors_tests.erl (hunk 1 above)
Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: Ah, good to know about Docs - thanks. I'll go ahead and make the change, as it's pretty minor and will indeed make the test more robust.

File: test/couchdb_mrview_cors_tests.erl (hunk 2 above)
Comment (kxepal) [inline test/couchdb_mrview_cors_tests.erl]: Basically, I think:
```
db_url(DbName) when is_binary(DbName) -> db_url(binary_to_list(DbName));
db_url(DbName) when is_list(DbName) -> host_url(cluster) ++ "/" ++ DbName.
host_url(Interface) -> "http://" ++ bind_address(Interface) ++ ":" ++ port(Interface).
bind_address(cluster) -> config:get("chttpd", "bind_address", "127.0.0.1");
bind_address(backdoor) -> config:get("httpd", "bind_address", "127.0.0.1").
port(cluster) -> integer_to_list(mochiweb_socket_server:get(chttpd, port));
port(backdoor) -> integer_to_list(mochiweb_socket_server:get(couch_httpd, port)).
```

Comment (jaydoane) [inline test/couchdb_mrview_cors_tests.erl]: Ah, this makes perfect sense now; thanks!

**Comment by jaydoane** [comment]
@kxepal, I believe the changes we discussed have been included in latest push.

**Comment by kxepal** [comment]
+1
